### ISSUE TITLE
Pix2Pix Code - Generator and Discriminator questions

### ISSUE THREAD

**Issue by pGit1** [eriklindernoren/Keras-GAN#36]
This repo is amazing!

Ive read the blogs and other post for pix2pix and your code seems to be the most clear but I have some questions about what exactly is going on:

1. In the "#Downsampling" portion of this code for the Generator, how exactly is any spatial resolution actually being lost (downsampled) if you fix the padding parameter equal to "same"? Technically arent the feature map sizes the exact same for d0 and d7 because of this regardless of stride and filter size? Padding equals same ensures this I thought. Since this part is puzzling I dont see how the "Upsampling" portion of the code (omitted here) is doing any actual upsampling back to d0 spatial resolution since we never got smaller than d0 to begin with. I think I am missing something really obvious here!

```
    def build_generator(self):
        """U-Net Generator"""

        def conv2d(layer_input, filters, f_size=4, bn=True):
            """Layers used during downsampling"""
            d = Conv2D(filters, kernel_size=f_size, strides=2, padding='same')(layer_input)
            d = LeakyReLU(alpha=0.2)(d)
            if bn:
                d = BatchNormalization(momentum=0.8)(d)
            return d

        def deconv2d(layer_input, skip_input, filters, f_size=4, dropout_rate=0):
            """Layers used during upsampling"""
            u = UpSampling2D(size=2)(layer_input)
            u = Conv2D(filters, kernel_size=f_size, strides=1, padding='same', activation='relu')(u)
            if dropout_rate:
                u = Dropout(dropout_rate)(u)
            u = BatchNormalization(momentum=0.8)(u)
            u = Concatenate()([u, skip_input])
            return u

        # Image input
        d0 = Input(shape=self.img_shape)

        # Downsampling
        d1 = conv2d(d0, self.gf, bn=False)
        d2 = conv2d(d1, self.gf*2)
        d3 = conv2d(d2, self.gf*4)
        d4 = conv2d(d3, self.gf*8)
        d5 = conv2d(d4, self.gf*8)
        d6 = conv2d(d5, self.gf*8)
        d7 = conv2d(d6, self.gf*8)

```

2. I am also having a hard time deciphering the code below concerning the discriminator. This patchGan idea is still pretty vague to me (not due to your code), wondering why it even matters but assuming I understood why it was neccessary I am still confused about the discriminator portion of the code since the "validity" variable would have the same spatial resolution as d1, which has the same spatial resolution as img_A/img_B for the same reason I outline above, also there is no explicit downsampling with the disriminator it seems, so there are two reasons why it seems H x W of validity should equal H x W of img_A/img_B, **however** when we go to train the discriminator we set "valid" and "fake" to a size much smaller than img_A or img_B, in our case img_A and img_B are supposed to be 256,256,3 but disc_patch has dimension 16,16,1. And it is this disc_patch dimension that acts as a target for the discriminator during training but if discrimnator recieves input of 256,256,3 it should return output at the same resolution resulting in 256,256,1 since I cant see anywhere in the code that would get the feature map size down to 16,16,1. Sorry for dumb questions but these are **burning** ones. I have more conceptual/fundemental questions about pix2pix in general but I cant ask them until I get past these roadblocks...

```

     # Calculate output shape of D (PatchGAN)
     patch = int(self.img_rows / 2**4)
     self.disc_patch = (patch, patch, 1) #has dim 16,16,1

    def build_discriminator(self):

        def d_layer(layer_input, filters, f_size=4, bn=True):
            """Discriminator layer"""
            d = Conv2D(filters, kernel_size=f_size, strides=2, padding='same')(layer_input)
            d = LeakyReLU(alpha=0.2)(d)
            if bn:
                d = BatchNormalization(momentum=0.8)(d)
            return d

        img_A = Input(shape=self.img_shape)
        img_B = Input(shape=self.img_shape)

        # Concatenate image and conditioning image by channels to produce input
        combined_imgs = Concatenate(axis=-1)([img_A, img_B])

        d1 = d_layer(combined_imgs, self.df, bn=False)
        d2 = d_layer(d1, self.df*2)
        d3 = d_layer(d2, self.df*4)
        d4 = d_layer(d3, self.df*8)

        validity = Conv2D(1, kernel_size=4, strides=1, padding='same')(d4)

        return Model([img_A, img_B], validity)

    def train(self, epochs, batch_size=1, sample_interval=50):

        start_time = datetime.datetime.now()

        # Adversarial loss ground truths
        valid = np.ones((batch_size,) + self.disc_patch)
        fake = np.zeros((batch_size,) + self.disc_patch)

        for epoch in range(epochs):
            for batch_i, (imgs_A, imgs_B) in enumerate(self.data_loader.load_batch(batch_size)):

                # ---------------------
                #  Train Discriminator
                # ---------------------

                # Condition on B and generate a translated version
                fake_A = self.generator.predict(imgs_B)

                # Train the discriminators (original images = real / generated = Fake)
                d_loss_real = self.discriminator.train_on_batch([imgs_A, imgs_B], valid)
                d_loss_fake = self.discriminator.train_on_batch([fake_A, imgs_B], fake)
                d_loss = 0.5 * np.add(d_loss_real, d_loss_fake)
```

THANKS FOR YOUR HELP!

**Comment by eriklindernoren** [comment]
Hi,

1) If you print the shape of the output of the last downsample layer in the generator you will see that the output has the shape `(?, 2, 2, 512)`. You can do this by adding `print (d7.shape)` in `Pix2Pix.build_generator`. By setting `padding='same'` padding will be applied to preserve the spatial resolution of the input given that `strides=1`. If you use the same setting for padding and set `strides=2` padding will be applied in the same way in the previous case, but as stride is set to 2 the spatial resolution of the output is downsampled by a factor of two.

2) As the input image is propogated through four downsample layers the output shape of the discriminator will be `(batch_size, img_height // 2**4, img_width // 2**4, 1) = (batch_size, 256 // 2**4, 256 // 2**4, 1)  = (batch_size, 16, 16, 1)`.

**Comment by pGit1** [comment]
Thanks for these answers!! 

1. All these years I've thought "same" was ensuring spatial resolution regardless of stride by automatically padding the the input feature maps with zeros...

But I just checked deeper and you are absolutely right "same" is ONLY maintains spatial when stride=1. Since I always use stride equals one, i was confused! Thanks for clarifying this...

2. Your answer to  number one answers this question of mine as well...

Thanks so much!!!!!!!

**Comment by eriklindernoren** [comment]
Awesome. No problem!

**Comment by pGit1** [comment]
I think I answered my own question about the Patch GAN idea by the way. Is the idea that rather than just letting D output a single number for the entire image as in the traditional sense, you instead force smaller sub patches of the image to be "believable" by D and as a result produce higher quality images from G. This seems to make a lot of sense to me. What is your recommendation for tuning the ouptut spatial resolution of D, is higher better or worse?

It looks like you make use of MSE losses as well is this due to the insights from Least Squares GANs paper (which I thought was fantastic by the way)? I agree with your choice of loss functions just wondering what your rationale is...

THIS CODE IS AMAZING and thanks to your efforts I think I have a 90%+ understanding of what is going with pix2pix. 

Concerning this Mode Collapse we always observe is there anything explicit in the code that I am missing that explicitly combats that? Ive heard of this term "mean distance between batch" but I dont know if that actually works well in practice.

Im closing this now but I would love your thoughts still!! 

THANKS AGAIN!!!!!!!!!!!!!
